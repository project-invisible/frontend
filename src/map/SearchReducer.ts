import { PointOfInterest } from "./../types/PointOfInterest";
import { CultureEntry } from "./../types/CultureEntry";

export const SEARCH_POI = "SEARCH_POI";
export const SEARCH_ENTRY = "SEARCH_ENTRY";
export const GET_POIS = "GET_POIS";
export const GET_ENTRIES = "GET_ENTRIES";
export const TOGGLE_SEARCH_LOADING = "TOGGLE_SEARCH_LOADING";
export const TOGGLE_ENTRY_SEARCH_LOADING = "TOGGLE_ENTRY_SEARCH_LOADING";
export const TOGGLE_RESET = "TOGGLE_RESET";
export const RESET_SEARCH = "RESET_SEARCH";

const initialState = {
  resetSearch: false,
  searchResults: [],
  searchEntryResults: [],
  allPois: [],
  allEntries: [],
  finishedEntryLoading: false,
  finishedFirstLoading: 0,
  finishedSearchLoading: false,
  finishedEntrySearchLoading: false
};

/**
 * Reducer
 */
const searchStore = (state = initialState, action) => {
  switch (action.type) {
    case SEARCH_POI:
      return {
        ...state,
        searchResults: action.searchResults,
        finishedSearchLoading: true
      };
    case SEARCH_ENTRY:
      return {
        ...state,
        searchEntryResults: action.searchEntriesResults,
        finishedEntrySearchLoading: true
      };
    case GET_POIS:
      const loadings = state.finishedFirstLoading + 1;
      return {
        ...state,
        allPois: action.pois,
        finishedFirstLoading: loadings
      };
    case GET_ENTRIES:
      return {
        ...state,
        finishedEntryLoading: true,
        allEntries: action.entries
      };
    case TOGGLE_SEARCH_LOADING:
      return {
        ...state,
        finishedSearchLoading: action.toggleLoading
      };
    case TOGGLE_RESET:
      return {
        ...state,
        resetSearch: action.toggleReset
      };
    case RESET_SEARCH:
      return {
        ...state,
        searchResults: [],
        searchEntryResults: [],
        finishedEntrySearchLoading: false,
        finishedSearchLoading: false,
        resetSearch: true
      };
    case TOGGLE_ENTRY_SEARCH_LOADING: 
      return {
        ...state,
        finishedEntrySearchLoading: action.toggleLoading
      }
  }
  return state;
};

/**
 * Actions
 */
export const searchPOI = (query: string) => async (dispatch, getState) => {
  try {
    const response = await fetch(
      `${process.env.REACT_APP_BACKEND_URL}/poi/search?query=${query}`,
      {
        method: "post",
        headers: { "Content-Type": "application/json" }
      }
    );
    const searchResults: Array<PointOfInterest> = await response.json();
    dispatch({
      type: SEARCH_POI,
      searchResults
    });
  } catch (error) {
    console.log("throwing Error", error);
    throw error;
  }
};

export const searchEntries = (query: string) => async (dispatch, getState) => {
  try {
    const response = await fetch(
      `${process.env.REACT_APP_BACKEND_URL}/entry/search?query=${query}`,
      {
        method: "post",
        headers: { "Content-Type": "application/json" }
      }
    );
    const searchEntriesResults: Array<CultureEntry> = await response.json();
    dispatch({
      type: SEARCH_ENTRY,
      searchEntriesResults
    });
  } catch (error) {
    console.log("throwing Error", error);
    throw error;
  }
};

export const toggleSearchLoading = (toggleLoading: boolean) => async (
  dispatch,
  getState
) => {
  try {
    dispatch({
      type: TOGGLE_SEARCH_LOADING,
      toggleLoading
    });
  } catch (error) {
    console.log("throwing Error", error);
    throw error;
  }
};

export const toggleEntrySearchLoading = (toggleLoading: boolean) => async (
  dispatch,
  getState
) => {
  try {
    dispatch({
      type: TOGGLE_ENTRY_SEARCH_LOADING,
      toggleLoading
    });
  } catch (error) {
    console.log("throwing Error", error);
    throw error;
  }
};

export const toggleResetSearch = (toggleReset: boolean) => async (
  dispatch,
  getState
) => {
  try {
    dispatch({
      type: TOGGLE_RESET,
      toggleReset
    });
  } catch (error) {
    console.log("throwing Error", error);
    throw error;
  }
};

export const resetSearch = () => async (dispatch, getState) => {
  try {
    dispatch({
      type: RESET_SEARCH
    });
  } catch (error) {
    console.log("throwing Error", error);
    throw error;
  }
};

export const getAllPOIs = () => async (dispatch, getState) => {
  try {
    const response = await fetch(`${process.env.REACT_APP_BACKEND_URL}/poi`);
    const body = await response.json();
    const pois = body;
    dispatch({
      type: GET_POIS,
      pois
    });
  } catch (error) {
    console.log("throwing Error", error);
    throw error;
  }
};

export const getAllEntries = () => async (dispatch, getState) => {
  try {
    const response = await fetch(`${process.env.REACT_APP_BACKEND_URL}/entry`);
    const body = await response.json();
    const entries = body;
    dispatch({
      type: GET_ENTRIES,
      entries
    });
  } catch (error) {
    console.log("throwing Error", error);
    throw error;
  }
};

/**
 * Store
 */
export default searchStore;
