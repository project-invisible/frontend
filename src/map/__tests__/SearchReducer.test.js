
import searchStore, { TOGGLE_ENTRY_SEARCH_LOADING, TOGGLE_RESET, GET_POIS, GET_ENTRIES, SEARCH_ENTRY, SEARCH_POI } from './../SearchReducer';
import {RESET_SEARCH} from '../SearchReducer';

describe('search reducer', () => {
  
it('should return the initial state', () => {
    expect(searchStore(undefined, {})).toMatchSnapshot();
  });

  it('reset search', () => {
      const expectedState = {
        searchResults: [],
        searchEntryResults: [],
        finishedEntrySearchLoading: false,
        finishedSearchLoading: false,
        resetSearch: true
      };
    expect(searchStore({}, {
      type: RESET_SEARCH,
    })).toEqual(expectedState);
  });

  it('toggle entry search loading', () => {
    const expectedState = {
        finishedEntrySearchLoading: true,
    };
  expect(searchStore({}, {
    type: TOGGLE_ENTRY_SEARCH_LOADING,
    toggleLoading: true
  })).toEqual(expectedState);
});


it('toggle reset', () => {
    const expectedState = {
        resetSearch: true
    };
  expect(searchStore({}, {
    type: TOGGLE_RESET,
    toggleReset: true
  })).toEqual(expectedState);
});

it('get pois', () => {
    const expectedState = {
        allPois:[1,2],
        finishedFirstLoading: 1
    };
  expect(searchStore({finishedFirstLoading: 0}, {
    type: GET_POIS,
    pois: [1,2]
  })).toEqual(expectedState);
});

it('get entries', () => {
    const expectedState = {
        finishedEntryLoading: true,
        allEntries: [1,2]
    };
  expect(searchStore({}, {
    type: GET_ENTRIES,
    entries: [1,2]
  })).toEqual(expectedState);
});

it('search entries', () => {
    const expectedState = {
        searchEntryResults: [1,2],
        finishedEntrySearchLoading: true
    };
  expect(searchStore({}, {
    type: SEARCH_ENTRY,
    searchEntriesResults: [1,2]
  })).toEqual(expectedState);
});

it('search poi', () => {
    const expectedState = {
        searchResults: [1,2],
        finishedSearchLoading: true
    };
  expect(searchStore({}, {
    type: SEARCH_POI,
    searchResults: [1,2]
  })).toEqual(expectedState);
});


});

