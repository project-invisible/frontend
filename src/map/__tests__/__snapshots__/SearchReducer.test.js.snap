// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search reducer should return the initial state 1`] = `
Object {
  "allEntries": Array [],
  "allPois": Array [],
  "finishedEntryLoading": false,
  "finishedEntrySearchLoading": false,
  "finishedFirstLoading": 0,
  "finishedSearchLoading": false,
  "resetSearch": false,
  "searchEntryResults": Array [],
  "searchResults": Array [],
}
`;
